File: app/src/main/java/com/bubelov/coins/ui/adapter/ExchangeRatesAdapter.java
package com.bubelov.coins.ui.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.bubelov.coins.R;
import com.bubelov.coins.model.ExchangeRate;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author Igor Bubelov
 */

public class ExchangeRatesAdapter extends RecyclerView.Adapter<ExchangeRatesAdapter.ExchangeRateViewHolder> {
    private final List<ExchangeRate> items = new ArrayList<>();

    @Override
    public ExchangeRateViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_exchange_rate, parent, false);
        return new ExchangeRateViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ExchangeRateViewHolder holder, int position) {
        ExchangeRate rate = items.get(position);
        holder.firstLetter.setText(rate.source().substring(0, 1));
        holder.exchangeName.setText(rate.source());
        holder.price.setText(String.format(Locale.US, "$%.2f", rate.rate()));
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public List<ExchangeRate> getItems() {
        return items;
    }

    static class ExchangeRateViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.first_letter)
        TextView firstLetter;

        @BindView(R.id.exchange_name)
        TextView exchangeName;

        @BindView(R.id.price)
        TextView price;

        ExchangeRateViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/com/bubelov/coins/ui/adapter/ExchangeRatesAdapter.kt
package com.bubelov.coins.ui.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.bubelov.coins.R
import com.bubelov.coins.ui.model.ExchangeRateQuery

import java.util.Locale

import kotlinx.android.synthetic.main.list_item_exchange_rate.view.*

/**
 * @author Igor Bubelov
 */

class ExchangeRatesAdapter : RecyclerView.Adapter<ExchangeRatesAdapter.ViewHolder>() {
    var items: List<ExchangeRateQuery> = emptyList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.list_item_exchange_rate, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(items[position])
    }

    override fun getItemCount(): Int {
        return items.size
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        fun bind(item: ExchangeRateQuery) {
            when(item) {
                is ExchangeRateQuery.Loading -> {
                    itemView.first_letter.text = item.source.substring(0, 1)
                    itemView.exchange_name.text = item.source
                    itemView.price.text = "fetching..."
                }
                is ExchangeRateQuery.ExchangeRate -> {
                    itemView.first_letter.text = item.source.substring(0, 1)
                    itemView.exchange_name.text = item.source
                    itemView.price.text = String.format(Locale.US, "$%.2f", item.rate)
                }
                is ExchangeRateQuery.Error -> {
                    itemView.first_letter.text = item.source.substring(0, 1)
                    itemView.exchange_name.text = item.source
                    itemView.price.text = "error"
                }
            }
        }
    }
}